core/dbt/cli/flags.py

```diff
@@ -89,102 +69,121 @@ def args_to_context(args: List[str]) -> Context:
+
+        # Ensure that any params sourced from the commandline are not present more than once.
+        # Click handles this exclusivity, but only at a per-subcommand level.
+        seen_params = []
+        for param in _get_params_by_source(ctx, ParameterSource.COMMANDLINE):
+            if param in seen_params:
+                raise BadOptionUsage(
+                    param.lower(),
+                    f"{param.lower()} was provided both before and after the subcommand, it can only be set either before or after.",
+                )
+            seen_params.append(param)
+
+        def _assign_params(
+            ctx: Context,
+            params_assigned_from_default: set,
+            deprecated_env_vars: Dict[str, Callable],
+        ):
             """Recursively adds all click params to flag object"""
             for param_name, param_value in ctx.params.items():
-                # TODO: this is to avoid duplicate params being defined in two places (version_check in run and cli)
-                # However this is a bit of a hack and we should find a better way to do this
-
                 # N.B. You have to use the base MRO method (object.__setattr__) to set attributes
                 # when using frozen dataclasses.
                 # https://docs.python.org/3/library/dataclasses.html#frozen-instances
-                if hasattr(self, param_name.upper()):
-                    if param_name in deprecated_env_vars:
-                        # param already set via its deprecated but still respected env var
+
+                # Handle deprecated env vars while still respecting old values
+                # e.g. DBT_NO_PRINT -> DBT_PRINT if DBT_NO_PRINT is set, it is
+                # respected over DBT_PRINT or --print.
+                new_name: Union[str, None] = None
+                if param_name in DEPRECATED_PARAMS:
+
+                    # Deprecated env vars can only be set via env var.
+                    # We use the deprecated option in click to serialize the value
+                    # from the env var string.
+                    param_source = ctx.get_parameter_source(param_name)
+                    if param_source == ParameterSource.DEFAULT:
                         continue
+                    elif param_source != ParameterSource.ENVIRONMENT:
+                        raise BadOptionUsage(
+                            param_name,
+                            "Deprecated parameters can only be set via environment variables",
+                        )
 
-                    if (
-                        param_name not in EXPECTED_DUPLICATE_PARAMS
-                        and ctx.get_parameter_source(param_name) != "kwargs"
-                    ):
+                    # Rename for clarity.
+                    dep_name = param_name
+                    new_name = DEPRECATED_PARAMS.get(dep_name)
+                    assert isinstance(new_name, str)
```

Comment:
I believe the `except IndexError` catch below should handle `dep_name` not existing in `DEPRECATED_PARAMS` and `new_name` being set to `None`.

Reply:
While that's probably true, mypy isn't happy unless we actually assert that it's a string. /shrug

Reply:
Would `new_name = DEPRECATED_PARAMS.get(dep_name, "")` satisfy this instead?

Reply:
No, but we can wrap it and raise a better looking exception. ✅